enhancements/machine-health-checking.md

```diff
@@ -0,0 +1,204 @@
+---
+title: machine-health-checking
+authors:
+  - @bison
+  - @ingvagabund
+  - @enxebre
+  - @cynepco3hahue
+reviewers:
+  - @derekwaynecarr
+  - @michaelgugino
+  - @beekhof
+  - @mrunalp
+approvers:
+  - @enxebre
+  - @bison
+  - @derekwaynecarr
+  - @michaelgugino
+
+creation-date: 2019-09-09
+last-updated: 2019-09-09
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Machine health checking
+
+## Release Signoff Checklist
+
+ - [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs
+
+
+## Summary
+
+Enable opt in for automated health checking and remediation of unhealhty nodes backed by machines.
+
+## Motivation
+
+Reduce administrative overhead to run a cluster and ability to respond to failures of machines.
+
+### Goals
+
+- Providing a default remediation strategy for public clouds - Machine deletion.
+
+- Allowing to define different unhealthy criterias for different pools of machines.
+
+- Allowing new and custom remediation strategies developed outside the machine-healthcheck controller, e.g baremetal reboot.
+
+- Providing administrator-defined short-circuiting of automated remediation.
+
+
+### Non-Goals
+
+- Coordinating individual remediation systems. This must be driven by each system however they see fit, e.g. by annotating affected nodes with state and transition time.
+
+- Recording long-term stable history of all health-check failures or remediations.
+
+## Proposal
+
+- As a customer I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated automatically.
+
+- As a cluster admin I want to define my unhealthy criteria as a list of node conditions https://github.com/kubernetes/node-problem-detector.
+
+- As a cluster admin I want to define different unhealthy criteria for pools of machines targeted for different workloads.
+
+- As a cluster admin I want to define different remediation strategies for different cloud environments. E.g deletion or reboot.
```

Comment:
I'm not against a reboot strategy (I don't think it really should be based on cloud environment, it should be based on machineset or similar resource), but it adds a lot of complexity and there are dependencies that don't exist.

1. machine-api currently doesn't have a 'reboot' mechanism
2. How many reboots should we attempt?  Do we eventually delete?
3. If we reboot a host and it doesn't come back, then what?  Need a statemachine to track all the things that could happen.  This should be articulated, and IMO, should be an entirely separate proposal.

Reply:
Agreed. This is stated in non-goals https://github.com/openshift/enhancements/pull/9/files#diff-53036ce0d8589cdbed15ed2e95076b54R59 this not about individual remediation systems but about serving as an integration point so the team working on "reboot" can flesh out their design. The only remediation strategy proposed here is "delete".